src/main.cpp

```diff
@@ -490,53 +546,179 @@ int main ( int argc, char** argv )
 
-    // per definition: if we are in "GUI" server mode and no directory server
-    // address is given, we use the default directory server address
-    if ( !bIsClient && bUseGUI && strCentralServer.isEmpty() )
-    {
-        strCentralServer = DEFAULT_SERVER_ADDRESS;
-    }
+            // if we are not headless, certain checks cannot be made, as the inifile state is not yet known
+            if ( !bUseGUI && strCentralServer.compare ( "localhost", Qt::CaseInsensitive ) != 0 && strCentralServer.compare ( "127.0.0.1" ) != 0 )
+            {
+                if ( !strServerListFileName.isEmpty() )
+                {
+                    qWarning() << "Server list persistence file will only take effect when running as a directory server.";
+                    strServerListFileName = "";
+                }
 
-    // adjust default port number for client: use different default port than the server since
-    // if the client is started before the server, the server would get a socket bind error
-    if ( bIsClient && !bCustomPortNumberGiven )
-    {
-        iPortNumber += 10; // increment by 10
+                if ( !strServerListFilter.isEmpty() )
+                {
+                    qWarning() << "Server list filter will only take effect when running as a directory server.";
+                    strServerListFileName = "";
+                }
+            }
+            else
+            {
+                if ( !strServerListFileName.isEmpty() )
+                {
+                    QFileInfo serverListFileInfo ( strServerListFileName );
+                    if ( !serverListFileInfo.exists() )
+                    {
+                        QFile strServerListFile ( strServerListFileName );
+                        if ( !strServerListFile.open ( QFile::OpenModeFlag::ReadWrite ) )
+                        {
+                            qWarning() << qUtf8Printable (
+                                QString ( "Cannot create %1 for reading and writing.  Please check permissions." ).arg ( strServerListFileName ) );
+                            strServerListFileName = "";
+                        }
+                    }
+                    else if ( !serverListFileInfo.isFile() )
+                    {
+                        qWarning() << qUtf8Printable (
+                            QString ( "Server list file %1 must be a plain file.  Please check the name." ).arg ( strServerListFileName ) );
+                        strServerListFileName = "";
+                    }
+                    else if ( !serverListFileInfo.isReadable() || !serverListFileInfo.isWritable() )
+                    {
+                        qWarning() << qUtf8Printable (
+                            QString ( "Server list file %1 must be readable and writeable.  Please check the permissions." )
+                                .arg ( strServerListFileName ) );
+                        strServerListFileName = "";
+                    }
+                }
+
```

Comment:
Hmm. It's a lot of new code in main.cpp then.

Reply:
Yep - you suggested that `--directoryfile` should only be allowed when running as a directory server.  That meant lots of new code.  As I was doing it for the one option, it felt necessary to be consistent, so I did it for as many I as I could.

Reply:
Yes. I know. Probably it's a good move then anyway. So the next steps would be another few tests and another person who reviews this.

Reply:
Currently CClientSettings and CServerSettings need, respectively, to be passed an instance of CClient and CServer.  Which means you have to create the Client or Server _before loading the inifile_.  The class in question configures itself according to its defaults or command line option overrides.  Then, if there's an inifile (i.e. not headless server), a whole load of other options get applied, changing how the class is configured.

This is why you _cannot_ know whether a server is a directory server if it's not headless when
- validating the command line options
- when creating the server instance

It's not until during inifile loading that you know.  That's after the directoryfile parameter has been set in the server.  Too late now!

All of this is confusing -- it leaves me totally confused every time I have to think about it, anyway.

Reply:
Thanks for the explanation. 

This seems to explain why my tests with the GUI didn't do what I thought they would do: I set up a GUI directory via CLI which resulted in a message saying it is a directory, but I couldn't register with it or have the register my server box checked. 

There must be a reason why Volker did it like it is at the moment?

Reply:
I think it was very much "that's the way it's always worked, why change it?".  Which is valid if you're trying to reduce the amount of change being made to something you have time to manage, which he was.